# PIPEseq

### A modular pipeline for VAMPseq

#### Pre-Configuration:
- conda (https://conda.io/projects/conda/en/latest/user-guide/install/index.html)
- bcl2fastq/2.20 (in cluster modules, technically bcl2fastq2)
- pear/0.9.11 (in cluster modules)
- fastqc/0.12.1 (in cluster modules)

#### Python Dependencies (will be installed during Setup)
- Python 3.10
- CutAdapt 4.9
- Snakemake 7.32.4
- CountESS (via pip)
  
#### Setup
- Clone PIPEseq into your directory
- In your terminal, navigate to the PIPEseq directory, `cd <directory_with_sequence_data>/PIPEseq`
- Create a Conda environment with yaml file `conda env create -f conda_env.yaml -n <env_name>`
- Activate the Conda environment `conda activate <env_name>`
- Copy your Samplesheet for bcl2fastq and barcode-variant map for VAMP-seq into the user_input folder
- Create an ini file from CountESS GUI if necessary
- Fill out at least base_sample_sheet, countess_sample_ini, and samplesheet_params in user_variables.yaml file, see User Variables section below

#### The Pipeline
- Rules:
    - clean_and_filter_samplesheet - modifies samplesheet to fit expected format for bcl2fastq
    - demux_and_pair - runs bcl2fastq and pear for demuxing and pairing reads
    - prep_fastqs_for_countess - trims fastqs (if necessary) and runs FastQC on all FASTQ files

#### User Variables
- The user_variables.yaml file can be modified with the following parameters, examples in example_yamls/:
**base_sample_sheet** - Name of Samplesheet CSV file
**countess_sample_ini** - Name of ini file for CountESS (TECHNICALLY not needed if only running step 1)
steps (OPTIONAL) - 1 (Only demux and pairing, creates FASTQs), 2 (Only CountESS, from existing FASTQs), leave blank for full pipeline **!!!THIS IS NOT READY YET!!!**
fastq_directory (OPTIONAL) - Location of FASTQs, if only running CountESS step **!!!THIS IS NOT READY YET!!!**
sample_filter (OPTIONAL) - If want to only run on certain samples (ex. F8)
**samplesheet_params** - The names of samplesheet columns if different from Illumina defaults
- **i7_index_id_column_name**
- **i7_index_sequence_column_name** - index
- **i5_index_id_column_name**
- **i5_index_sequence_column_name** - index2
**cutadapt_trim_boolean** - Whether or not trimming is necessary
cutadapt_params (OPTIONAL, including all below) - For use if trimming with CutAdapt
- error
- f_amp_min_overlap
- r_amp_min_overlap
- f_amp_primer_column_name - upstream flanking sequence
- r_amp_primer_column_name - downstream flanking sequence
- target_length_column_name
**bold** = needs to be filled in by user

#### Expected Folder Structure
- *<directory_with_raw_data>/
    - PIPEseq/
        - example_yamls/
        - **run-.../
        - user_input/
            - *<samplesheet_file>.csv
            - *<countess_file>.ini
        - conda_env.yaml
        - README.md
        - run.sh
        - Snakefile
        - &user_variables.yaml

*=user file (or folder)
**=autogenerated
&=modify parameters

